CS1585	Member modifier 'replace' must precede the member type and name

**Version Used**:

Bug seen in Roslyn Future  20160520.2 

**Steps to Reproduce**:
Copy and paste the below code

``` C#
using System.Text.RegularExpressions;

namespace TestCS1585
{
    public class Repro
    {
        public static string replace(string input, string regexp, string flags, string replacement)
        {

            RegexOptions options = RegexOptions.ECMAScript;

            if (flags.IndexOf("i") != -1)
            {
                options |= RegexOptions.IgnoreCase;
            }

            Regex regex = new Regex(regexp, options);

            if (flags.IndexOf("g") != -1)
            {
                return regex.Replace(input, replacement, 1);
            }
            else
            {
                return regex.Replace(input, replacement);
            }
        }
    }
}
```

**Expected Behavior**:

No errors

**Actual Behavior**:

Repro.cs(12,30,12,37): error CS1585: Member modifier 'replace' must precede the member type and name
Repro.cs(12,30,12,37): error CS1520: Method must have a return type

--- a/src/Compilers/CSharp/Portable/Parser/LanguageParser.cs
+++ b/src/Compilers/CSharp/Portable/Parser/LanguageParser.cs
@@ -2465,6 +2465,7 @@ tryAgain:
                 if (GetModifier(this.CurrentToken) != SyntaxModifier.None &&
                     this.CurrentToken.ContextualKind != SyntaxKind.PartialKeyword &&
                     this.CurrentToken.ContextualKind != SyntaxKind.AsyncKeyword &&
+                    this.CurrentToken.ContextualKind != SyntaxKind.ReplaceKeyword &&
                     IsComplete(type))
                 {
                     var misplacedModifier = this.CurrentToken;
@@ -2586,8 +2587,8 @@ parse_member_name:;
             }
         }
 
-        // if the modifiers do not contain async and the type is the identifier "async", then
-        // add async to the modifiers and assign a new type from the identifierOrThisOpt and the
+        // if the modifiers do not contain async or replace and the type is the identifier "async" or "replace", then
+        // add that identifier to the modifiers and assign a new type from the identifierOrThisOpt and the
         // type parameter list
         private bool ReconsiderTypeAsAsyncModifier(
             ref SyntaxListBuilder modifiers,
@@ -2596,11 +2597,18 @@ parse_member_name:;
             SyntaxToken identifierOrThisOpt,
             TypeParameterListSyntax typeParameterListOpt)
         {
-            if (modifiers.Any(SyntaxKind.AsyncKeyword)) return false;
             if (type.Kind != SyntaxKind.IdentifierName) return false;
-            if ((((IdentifierNameSyntax)type).Identifier).ContextualKind != SyntaxKind.AsyncKeyword) return false;
             if (identifierOrThisOpt.Kind != SyntaxKind.IdentifierToken) return false;
-            modifiers.Add(ConvertToKeyword(((IdentifierNameSyntax)type).Identifier));
+
+            var identifier = ((IdentifierNameSyntax)type).Identifier;
+            var contextualKind = identifier.ContextualKind;
+            if ((contextualKind != SyntaxKind.AsyncKeyword && contextualKind != SyntaxKind.ReplaceKeyword) ||
+                modifiers.Any(contextualKind))
+            {
+                return false;
+            }
+
+            modifiers.Add(ConvertToKeyword(identifier));
             SimpleNameSyntax newType = typeParameterListOpt == null
                 ? (SimpleNameSyntax)_syntaxFactory.IdentifierName(identifierOrThisOpt)
                 : _syntaxFactory.GenericName(identifierOrThisOpt, TypeArgumentFromTypeParameters(typeParameterListOpt));

--- a/src/Compilers/CSharp/Test/Symbol/Symbols/ReplaceOriginalTests.cs
+++ b/src/Compilers/CSharp/Test/Symbol/Symbols/ReplaceOriginalTests.cs
@@ -20,7 +20,6 @@ namespace Microsoft.CodeAnalysis.CSharp.UnitTests.Symbols
         {
             return CreateCompilationWithMscorlib(text, options: options, parseOptions: TestOptions.Regular.WithReplaceFeature());
         }
-
         
         [WorkItem(11123, "https://github.com/dotnet/roslyn/issues/11123")]
         [Fact]
@@ -57,7 +56,6 @@ partial struct S
             compilation.VerifyDiagnostics();
         }
 
-        
         [Fact]
         public void ExpressionBodiedMembers()
         {
@@ -83,7 +81,6 @@ partial struct S
   IL_0008:  ret
 }");
         }
-
         
         [Fact]
         public void Fields()
@@ -124,7 +121,6 @@ partial struct S
                 Diagnostic(ErrorCode.WRN_UnreferencedField, "F").WithArguments("C.F").WithLocation(3, 19));
         }
 
-        
         [Fact]
         public void Types()
         {
@@ -149,7 +145,6 @@ class B
                 //     replace class C { }
                 Diagnostic(ErrorCode.ERR_DuplicateNameInClass, "C").WithArguments("B", "C").WithLocation(7, 19));
         }
-
         
         [Fact]
         public void Interface()
@@ -177,7 +172,6 @@ class B
                 //     replace object P { get; }
                 Diagnostic(ErrorCode.ERR_DuplicateNameInClass, "P").WithArguments("I", "P").WithLocation(6, 20));
         }
-
         
         [Fact]
         public void Accessors()
@@ -202,7 +196,6 @@ class C
                 //     event EventHandler E { replace add { } remove { } }
                 Diagnostic(ErrorCode.ERR_AddOrRemoveExpected, "replace").WithLocation(6, 28));
         }
-
         
         [Fact]
         public void Constructors()
@@ -241,7 +234,6 @@ struct S
                 //     replace internal S(object o) { }
                 Diagnostic(ErrorCode.ERR_MemberAlreadyExists, "S").WithArguments(".ctor", "S").WithLocation(11, 22));
         }
-
         
         [Fact]
         public void DefaultConstructors()
@@ -267,7 +259,6 @@ struct S
                 //     replace public S() { }
                 Diagnostic(ErrorCode.ERR_StructsCantContainDefaultConstructor, "S").WithLocation(7, 20));
         }
-
         
         [Fact]
         public void ImplicitConstructors()
@@ -290,7 +281,6 @@ struct S
                 //     replace public C() { }
                 Diagnostic(ErrorCode.ERR_BadMemberFlag, "C").WithArguments("replace").WithLocation(7, 20));
         }
-
         
         [Fact]
         public void Destructors()
@@ -310,7 +300,6 @@ struct S
                 //     replace ~C() { }
                 Diagnostic(ErrorCode.ERR_MemberAlreadyExists, "C").WithArguments("~C", "C").WithLocation(4, 14));
         }
-
         
         [Fact]
         public void FinalizeMethod()
@@ -331,7 +320,6 @@ struct S
                 Diagnostic(ErrorCode.ERR_OverrideFinalizeDeprecated, "Finalize").WithLocation(4, 37));
         }
 
-        
         [WorkItem(11122, "https://github.com/dotnet/roslyn/issues/11122")]
         [Fact]
         public void ExplicitImplementation()
@@ -382,7 +370,6 @@ class C : I
                 //         set { original[index] = value; }
                 Diagnostic(ErrorCode.ERR_NoOriginalMember, "original").WithArguments("C.this[int]").WithLocation(22, 15));
         }
-
         
         [Fact]
         public void MultipleReplaces()
@@ -448,7 +435,6 @@ class C
                 //     event EventHandler E;
                 Diagnostic(ErrorCode.WRN_UnreferencedEvent, "E").WithArguments("C.E").WithLocation(20, 24));
         }
-
         
         [WorkItem(11124, "https://github.com/dotnet/roslyn/issues/11124")]
         [Fact]
@@ -511,7 +497,6 @@ class C
             }
         }
 
-        
         [WorkItem(11125, "https://github.com/dotnet/roslyn/issues/11125")]
         [Fact]
         public void DifferentSignatures()
@@ -544,7 +529,6 @@ class C
             var compilation = CreateCompilationWithMscorlib(source);
             compilation.VerifyDiagnostics(); 
         }
-
         
         [WorkItem(11126, "https://github.com/dotnet/roslyn/issues/11126")]
         [Fact]
@@ -562,7 +546,6 @@ class C
             compilation.VerifyDiagnostics();
         }
 
-        
         [WorkItem(11127, "https://github.com/dotnet/roslyn/issues/11127")]
         [Fact]
         public void Abstract()
@@ -578,7 +561,6 @@ class C
             var compilation = CreateCompilationWithMscorlib(source);
             compilation.VerifyDiagnostics(); 
         }
-
         
         [Fact]
         public void Virtual()
@@ -654,7 +636,6 @@ class C
                 }
             }
         }
-
         
         [Fact]
         public void Overridden()
@@ -684,7 +665,6 @@ class B : A
   IL_0006:  ret
 }");
         }
-
         
         [Fact]
         public void Attributes()
@@ -703,7 +683,6 @@ class C
             compilation.VerifyDiagnostics();
         }
 
-        
         [WorkItem(11129, "https://github.com/dotnet/roslyn/issues/11129")]
         [Fact]
         public void DifferentAttributes()
@@ -726,7 +705,6 @@ class C
             compilation.VerifyDiagnostics(); 
         }
 
-        
         [WorkItem(11130, "https://github.com/dotnet/roslyn/issues/11130")]
         [Fact]
         public void Modifiers()
@@ -734,7 +712,6 @@ class C
             // Test extern, partial, unsafe, async
         }
 
-        
         [Fact]
         public void MissingSet()
         {
@@ -750,7 +727,6 @@ class C
                 //     replace object P { get { original = null; return null; } }
                 Diagnostic(ErrorCode.ERR_AssgReadonlyProp, "original").WithArguments("C.P").WithLocation(4, 30));
         }
-
         
         [Fact]
         public void MissingGet()
@@ -767,7 +743,6 @@ class C
                 //     replace int P { set { original += value; } }
                 Diagnostic(ErrorCode.ERR_PropertyLacksGet, "original").WithArguments("C.P").WithLocation(4, 27));
         }
-
         
         [Fact]
         public void OriginalAsDelegate()
@@ -790,7 +765,6 @@ class C
             compilation.VerifyDiagnostics();
         }
 
-        
         [Fact]
         public void OriginalInLambda()
         {
@@ -820,7 +794,6 @@ class C
             compilation.VerifyDiagnostics();
             CompileAndVerify(compilation, expectedOutput: @"21");
         }
-
         
         [WorkItem(11131, "https://github.com/dotnet/roslyn/issues/11131")]
         [Fact]
@@ -837,7 +810,6 @@ class C
             var compilation = CreateCompilationWithMscorlib(source);
             compilation.VerifyDiagnostics();
         }
-
         
         [WorkItem(11131, "https://github.com/dotnet/roslyn/issues/11131")]
         [Fact]
@@ -854,7 +826,6 @@ class C
             var compilation = CreateCompilationWithMscorlib(source);
             compilation.VerifyDiagnostics(); 
         }
-
         
         [WorkItem(11131, "https://github.com/dotnet/roslyn/issues/11131")]
         [Fact]
@@ -871,7 +842,6 @@ class C
             var compilation = CreateCompilationWithMscorlib(source);
             compilation.VerifyDiagnostics(); 
         }
-
         
         [Fact]
         public void ReplaceWithAutoProperty()
@@ -892,7 +862,6 @@ class C
             compilation.VerifyDiagnostics();
         }
 
-        
         [Fact]
         public void ReplaceWithFieldLikeEvent()
         {
@@ -917,7 +886,6 @@ class C
                 //     replace event EventHandler E;
                 Diagnostic(ErrorCode.WRN_UnreferencedEvent, "E").WithArguments("C.E").WithLocation(6, 32));
         }
-
         
         [Fact]
         public void ParamsArray()
@@ -942,7 +910,6 @@ class C
         /// Replaced generic method. Original reference
         /// must use type arguments from replace method.
         /// </summary>
-        
         [Fact]
         public void GenericMethod()
         {
@@ -955,7 +922,6 @@ class C
             var compilation = CreateCompilationWithMscorlib(source);
             compilation.VerifyDiagnostics();
         }
-
         
         [Fact]
         public void OriginalMethodDifferentTypeArguments()
@@ -979,7 +945,6 @@ class C
                 Diagnostic(ErrorCode.ERR_BadArgType, "2").WithArguments("2", "int", "U").WithLocation(9, 21));
         }
 
-        
         [WorkItem(11132, "https://github.com/dotnet/roslyn/issues/11132")]
         [Fact]
         public void OriginalMethodExplicitTypeArguments()
@@ -1011,7 +976,6 @@ class C
                 //         original<T>(default(T));
                 Diagnostic(ErrorCode.ERR_IllegalStatement, "original<T>(default(T))").WithLocation(8, 9));
         }
-
         
         [Fact]
         public void PartialReplace()
@@ -1063,7 +1027,6 @@ class C
                 Diagnostic(ErrorCode.ERR_MemberAlreadyExists, "M4").WithArguments("M4", "C").WithLocation(18, 26));
         }
 
-        
         [WorkItem(11117, "https://github.com/dotnet/roslyn/issues/11117")]
         [Fact]
         public void ReplacePartial()
@@ -1103,7 +1066,52 @@ class C
                 Diagnostic(ErrorCode.ERR_NoOriginalMember, "original").WithArguments("C.M2()").WithLocation(13, 9));
         }
 
-        
+        [Fact]
+        public void MethodNamedReplace()
+        {
+            var source =
+@"using System;
+class C
+{
+    void replace(bool b)
+    {
+        Console.Write(1);
+    }
+    replace void replace(bool b)
+    {
+        Console.Write(2);
+        original(b);
+        if (b)
+        {
+            replace(false);
+        }
+    }
+    static void Main()
+    {
+        (new C()).replace(true);
+    }
+}";
+            var compilation = CreateCompilationWithMscorlib(source, options: TestOptions.ReleaseExe);
+            compilation.VerifyDiagnostics();
+            var verifier = CompileAndVerify(compilation, expectedOutput: @"2121");
+            VerifyIL(verifier, GetMethodByName(verifier.TestData, "C", "replace"),
+@"{
+  // Code size       24 (0x18)
+  .maxstack  2
+  IL_0000:  ldc.i4.2
+  IL_0001:  call       ""void System.Console.Write(int)""
+  IL_0006:  ldarg.0
+  IL_0007:  ldarg.1
+  IL_0008:  call       ""void C.<replace>v__0(bool)""
+  IL_000d:  ldarg.1
+  IL_000e:  brfalse.s  IL_0017
+  IL_0010:  ldarg.0
+  IL_0011:  ldc.i4.0
+  IL_0012:  call       ""void C.replace(bool)""
+  IL_0017:  ret
+}");
+        }
+
         [Fact]
         public void MethodNamedOriginal()
         {
@@ -1154,7 +1162,6 @@ class C
 }");
         }
 
-        
         [Fact]
         public void EntryPoint()
         {
@@ -1200,14 +1207,12 @@ class C
 }");
         }
 
-        
         [WorkItem(11134, "https://github.com/dotnet/roslyn/issues/11134")]
         [Fact]
         public void UseSiteDiagnostics()
         {
             // Can we test useSiteDiagnostics in BindInvocationExpression?
         }
-
         
         [Fact]
         public void LookupReplace()

--- a/src/Compilers/CSharp/Test/Syntax/CSharpCompilerSyntaxTest.csproj
+++ b/src/Compilers/CSharp/Test/Syntax/CSharpCompilerSyntaxTest.csproj
@@ -91,6 +91,7 @@
     <Compile Include="Parsing\DeclarationParsingTests.cs" />
     <Compile Include="Parsing\ExpressionParsingTests.cs" />
     <Compile Include="Parsing\LocalFunctionParsingTests.cs" />
+    <Compile Include="Parsing\MoreReplaceParsingTests.cs" />
     <Compile Include="Parsing\ParserRegressionTests.cs" />
     <Compile Include="Parsing\ReplaceParsingTests.cs" />
     <Compile Include="Parsing\ScriptParsingTests.cs" />

--- a/src/Compilers/CSharp/Test/Syntax/Parsing/MoreReplaceParsingTests.cs
+++ b/src/Compilers/CSharp/Test/Syntax/Parsing/MoreReplaceParsingTests.cs
@@ -1,0 +1,765 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using Microsoft.CodeAnalysis.CSharp.Test.Utilities;
+using Microsoft.CodeAnalysis.Test.Utilities;
+using Roslyn.Test.Utilities;
+using Xunit;
+
+namespace Microsoft.CodeAnalysis.CSharp.UnitTests
+{
+    // Adapted from AsyncParsingTests.
+    [CompilerTrait(CompilerFeature.SourceGenerators)]
+    public class MoreReplaceParsingTests : ParsingTests
+    {
+        protected override SyntaxTree ParseTree(string text, CSharpParseOptions options)
+        {
+            return SyntaxFactory.ParseSyntaxTree(text, options: (options ?? TestOptions.Regular).WithReplaceFeature());
+        }
+
+        protected override CSharpSyntaxNode ParseNode(string text, CSharpParseOptions options)
+        {
+            throw new System.NotImplementedException();
+        }
+
+        [Fact]
+        public void MethodCalledReplace()
+        {
+            UsingTree(@"
+class C
+{
+    void replace() { }
+}
+");
+            N(SyntaxKind.CompilationUnit);
+            {
+                N(SyntaxKind.ClassDeclaration);
+                {
+                    N(SyntaxKind.ClassKeyword);
+                    N(SyntaxKind.IdentifierToken);
+                    N(SyntaxKind.OpenBraceToken);
+                    N(SyntaxKind.MethodDeclaration);
+                    {
+                        N(SyntaxKind.PredefinedType);
+                        {
+                            N(SyntaxKind.VoidKeyword);
+                        }
+                        N(SyntaxKind.IdentifierToken);
+                        N(SyntaxKind.ParameterList);
+                        {
+                            N(SyntaxKind.OpenParenToken);
+                            N(SyntaxKind.CloseParenToken);
+                        }
+                        N(SyntaxKind.Block);
+                        {
+                            N(SyntaxKind.OpenBraceToken);
+                            N(SyntaxKind.CloseBraceToken);
+                        }
+                    }
+                    N(SyntaxKind.CloseBraceToken);
+                }
+                N(SyntaxKind.EndOfFileToken);
+            }
+        }
+
+        [Fact]
+        public void MethodReturningReplace()
+        {
+            UsingTree(@"
+class C
+{
+    replace M() { }
+}
+");
+            N(SyntaxKind.CompilationUnit);
+            {
+                N(SyntaxKind.ClassDeclaration);
+                {
+                    N(SyntaxKind.ClassKeyword);
+                    N(SyntaxKind.IdentifierToken);
+                    N(SyntaxKind.OpenBraceToken);
+                    N(SyntaxKind.MethodDeclaration);
+                    {
+                        N(SyntaxKind.IdentifierName);
+                        {
+                            N(SyntaxKind.IdentifierToken);
+                        }
+                        N(SyntaxKind.IdentifierToken);
+                        N(SyntaxKind.ParameterList);
+                        {
+                            N(SyntaxKind.OpenParenToken);
+                            N(SyntaxKind.CloseParenToken);
+                        }
+                        N(SyntaxKind.Block);
+                        {
+                            N(SyntaxKind.OpenBraceToken);
+                            N(SyntaxKind.CloseBraceToken);
+                        }
+                    }
+                    N(SyntaxKind.CloseBraceToken);
+                }
+                N(SyntaxKind.EndOfFileToken);
+            }
+        }
+
+        [Fact]
+        public void MethodReplaceReplace()
+        {
+            UsingTree(@"
+class C
+{
+    replace replace() { }
+}
+");
+            N(SyntaxKind.CompilationUnit);
+            {
+                N(SyntaxKind.ClassDeclaration);
+                {
+                    N(SyntaxKind.ClassKeyword);
+                    N(SyntaxKind.IdentifierToken);
+                    N(SyntaxKind.OpenBraceToken);
+                    N(SyntaxKind.MethodDeclaration);
+                    {
+                        N(SyntaxKind.IdentifierName);
+                        {
+                            N(SyntaxKind.IdentifierToken);
+                        }
+                        N(SyntaxKind.IdentifierToken);
+                        N(SyntaxKind.ParameterList);
+                        {
+                            N(SyntaxKind.OpenParenToken);
+                            N(SyntaxKind.CloseParenToken);
+                        }
+                        N(SyntaxKind.Block);
+                        {
+                            N(SyntaxKind.OpenBraceToken);
+                            N(SyntaxKind.CloseBraceToken);
+                        }
+                    }
+                    N(SyntaxKind.CloseBraceToken);
+                }
+                N(SyntaxKind.EndOfFileToken);
+            }
+        }
+
+        [Fact]
+        public void MethodReplaceReplaceReplace()
+        {
+            UsingTree(@"
+class C
+{
+    replace replace replace() { }
+}
+");
+            N(SyntaxKind.CompilationUnit);
+            {
+                N(SyntaxKind.ClassDeclaration);
+                {
+                    N(SyntaxKind.ClassKeyword);
+                    N(SyntaxKind.IdentifierToken);
+                    N(SyntaxKind.OpenBraceToken);
+                    N(SyntaxKind.MethodDeclaration);
+                    {
+                        N(SyntaxKind.ReplaceKeyword);
+                        N(SyntaxKind.IdentifierName);
+                        {
+                            N(SyntaxKind.IdentifierToken);
+                        }
+                        N(SyntaxKind.IdentifierToken);
+                        N(SyntaxKind.ParameterList);
+                        {
+                            N(SyntaxKind.OpenParenToken);
+                            N(SyntaxKind.CloseParenToken);
+                        }
+                        N(SyntaxKind.Block);
+                        {
+                            N(SyntaxKind.OpenBraceToken);
+                            N(SyntaxKind.CloseBraceToken);
+                        }
+                    }
+                    N(SyntaxKind.CloseBraceToken);
+                }
+                N(SyntaxKind.EndOfFileToken);
+            }
+        }
+
+        [Fact]
+        public void IncompleteReplace()
+        {
+            UsingTree(@"
+class C
+{
+    replace
+");
+            N(SyntaxKind.CompilationUnit);
+            {
+                N(SyntaxKind.ClassDeclaration);
+                {
+                    N(SyntaxKind.ClassKeyword);
+                    N(SyntaxKind.IdentifierToken);
+                    N(SyntaxKind.OpenBraceToken);
+                    N(SyntaxKind.IncompleteMember);
+                    {
+                        N(SyntaxKind.IdentifierName);
+                        {
+                            N(SyntaxKind.IdentifierToken);
+                        }
+                    }
+                    M(SyntaxKind.CloseBraceToken);
+                }
+                N(SyntaxKind.EndOfFileToken);
+            }
+        }
+
+        [Fact]
+        public void IncompleteReplaceReplace()
+        {
+            UsingTree(@"
+class C
+{
+    replace replace
+");
+            N(SyntaxKind.CompilationUnit);
+            {
+                N(SyntaxKind.ClassDeclaration);
+                {
+                    N(SyntaxKind.ClassKeyword);
+                    N(SyntaxKind.IdentifierToken);
+                    N(SyntaxKind.OpenBraceToken);
+                    N(SyntaxKind.FieldDeclaration);
+                    {
+                        N(SyntaxKind.VariableDeclaration);
+                        {
+                            N(SyntaxKind.IdentifierName);
+                            {
+                                N(SyntaxKind.IdentifierToken);
+                            }
+                            N(SyntaxKind.VariableDeclarator);
+                            {
+                                N(SyntaxKind.IdentifierToken);
+                            }
+                        }
+                        M(SyntaxKind.SemicolonToken);
+                    }
+                    M(SyntaxKind.CloseBraceToken);
+                }
+                N(SyntaxKind.EndOfFileToken);
+            }
+        }
+
+        [Fact]
+        public void IncompleteReplaceReplaceReplace()
+        {
+            UsingTree(@"
+class C
+{
+    replace replace replace
+");
+            N(SyntaxKind.CompilationUnit);
+            {
+                N(SyntaxKind.ClassDeclaration);
+                {
+                    N(SyntaxKind.ClassKeyword);
+                    N(SyntaxKind.IdentifierToken);
+                    N(SyntaxKind.OpenBraceToken);
+                    N(SyntaxKind.FieldDeclaration);
+                    {
+                        N(SyntaxKind.ReplaceKeyword);
+                        N(SyntaxKind.VariableDeclaration);
+                        {
+                            N(SyntaxKind.IdentifierName);
+                            {
+                                N(SyntaxKind.IdentifierToken);
+                            }
+                            N(SyntaxKind.VariableDeclarator);
+                            {
+                                N(SyntaxKind.IdentifierToken);
+                            }
+                        }
+                        M(SyntaxKind.SemicolonToken);
+                    }
+                    M(SyntaxKind.CloseBraceToken);
+                }
+                N(SyntaxKind.EndOfFileToken);
+            }
+        }
+
+        [WorkItem(609912, "http://vstfdevdiv:8080/DevDiv2/DevDiv/_workitems/edit/609912")]
+        [Fact]
+        public void IncompleteReplaceMember01()
+        {
+            // when parsing an incomplete member, treat 'replace' as a modifier if it makes sense
+            UsingTree(@"
+class C
+{
+    replace Task<
+}");
+            N(SyntaxKind.CompilationUnit);
+            {
+                N(SyntaxKind.ClassDeclaration);
+                {
+                    N(SyntaxKind.ClassKeyword);
+                    N(SyntaxKind.IdentifierToken);
+                    N(SyntaxKind.OpenBraceToken);
+                    N(SyntaxKind.IncompleteMember);
+                    {
+                        N(SyntaxKind.ReplaceKeyword);
+                        N(SyntaxKind.GenericName);
+                        {
+                            N(SyntaxKind.IdentifierToken);
+                            N(SyntaxKind.TypeArgumentList);
+                            {
+                                N(SyntaxKind.LessThanToken);
+                                N(SyntaxKind.IdentifierName);
+                                {
+                                    N(SyntaxKind.IdentifierToken);
+                                }
+                                M(SyntaxKind.GreaterThanToken);
+                            }
+                        }
+                    }
+                    N(SyntaxKind.CloseBraceToken);
+                }
+                N(SyntaxKind.EndOfFileToken);
+            }
+        }
+
+        [WorkItem(609912, "http://vstfdevdiv:8080/DevDiv2/DevDiv/_workitems/edit/609912")]
+        [Fact]
+        public void IncompleteReplaceMember02()
+        {
+            // when parsing an incomplete member, treat 'replace' as a modifier if it makes sense
+            UsingTree(@"
+class C
+{
+    replace Tasks.Task<
+}");
+            N(SyntaxKind.CompilationUnit);
+            {
+                N(SyntaxKind.ClassDeclaration);
+                {
+                    N(SyntaxKind.ClassKeyword);
+                    N(SyntaxKind.IdentifierToken);
+                    N(SyntaxKind.OpenBraceToken);
+                    N(SyntaxKind.IncompleteMember);
+                    {
+                        N(SyntaxKind.ReplaceKeyword);
+                        N(SyntaxKind.QualifiedName);
+                        {
+                            N(SyntaxKind.IdentifierName);
+                            {
+                                N(SyntaxKind.IdentifierToken);
+                            }
+                            N(SyntaxKind.DotToken);
+                            N(SyntaxKind.GenericName);
+                            {
+                                N(SyntaxKind.IdentifierToken);
+                                N(SyntaxKind.TypeArgumentList);
+                                {
+                                    N(SyntaxKind.LessThanToken);
+                                    N(SyntaxKind.IdentifierName);
+                                    {
+                                        N(SyntaxKind.IdentifierToken);
+                                    }
+                                    M(SyntaxKind.GreaterThanToken);
+                                }
+                            }
+                        }
+                    }
+                    N(SyntaxKind.CloseBraceToken);
+                }
+                N(SyntaxKind.EndOfFileToken);
+            }
+        }
+
+        [WorkItem(609912, "http://vstfdevdiv:8080/DevDiv2/DevDiv/_workitems/edit/609912")]
+        [Fact]
+        public void IncompleteReplaceMember03()
+        {
+            // when parsing an incomplete member, treat 'replace' as a modifier if it makes sense
+            UsingTree(@"
+class C
+{
+    static replace Tasks.Task<
+}");
+            N(SyntaxKind.CompilationUnit);
+            {
+                N(SyntaxKind.ClassDeclaration);
+                {
+                    N(SyntaxKind.ClassKeyword);
+                    N(SyntaxKind.IdentifierToken);
+                    N(SyntaxKind.OpenBraceToken);
+                    N(SyntaxKind.IncompleteMember);
+                    {
+                        N(SyntaxKind.StaticKeyword);
+                        N(SyntaxKind.ReplaceKeyword);
+                        N(SyntaxKind.QualifiedName);
+                        {
+                            N(SyntaxKind.IdentifierName);
+                            {
+                                N(SyntaxKind.IdentifierToken);
+                            }
+                            N(SyntaxKind.DotToken);
+                            N(SyntaxKind.GenericName);
+                            {
+                                N(SyntaxKind.IdentifierToken);
+                                N(SyntaxKind.TypeArgumentList);
+                                {
+                                    N(SyntaxKind.LessThanToken);
+                                    N(SyntaxKind.IdentifierName);
+                                    {
+                                        N(SyntaxKind.IdentifierToken);
+                                    }
+                                    M(SyntaxKind.GreaterThanToken);
+                                }
+                            }
+                        }
+                    }
+                    N(SyntaxKind.CloseBraceToken);
+                }
+                N(SyntaxKind.EndOfFileToken);
+            }
+        }
+
+        [WorkItem(609912, "http://vstfdevdiv:8080/DevDiv2/DevDiv/_workitems/edit/609912")]
+        [Fact]
+        public void IncompleteReplaceMember04()
+        {
+            // when parsing an incomplete member, treat 'replace' as a modifier if it makes sense
+            // negative case
+            UsingTree(@"
+class C
+{
+    replace operator+
+}");
+            N(SyntaxKind.CompilationUnit);
+            {
+                N(SyntaxKind.ClassDeclaration);
+                {
+                    N(SyntaxKind.ClassKeyword);
+                    N(SyntaxKind.IdentifierToken);
+                    N(SyntaxKind.OpenBraceToken);
+                    N(SyntaxKind.OperatorDeclaration);
+                    {
+                        N(SyntaxKind.IdentifierName);
+                        {
+                            N(SyntaxKind.IdentifierToken); // replace
+                        }
+                        N(SyntaxKind.OperatorKeyword);
+                        N(SyntaxKind.PlusToken);
+                        M(SyntaxKind.ParameterList);
+                        {
+                            M(SyntaxKind.OpenParenToken);
+                            M(SyntaxKind.CloseParenToken);
+                        }
+                        M(SyntaxKind.SemicolonToken);
+                    }
+                    N(SyntaxKind.CloseBraceToken);
+                }
+                N(SyntaxKind.EndOfFileToken);
+            }
+        }
+
+        [WorkItem(609912, "http://vstfdevdiv:8080/DevDiv2/DevDiv/_workitems/edit/609912")]
+        [Fact]
+        public void IncompleteReplaceMember05()
+        {
+            // when parsing an incomplete member, treat 'replace' as a modifier if it makes sense
+            // negative case
+            UsingTree(@"
+class C
+{
+    replace Task<T>
+}");
+            N(SyntaxKind.CompilationUnit);
+            {
+                N(SyntaxKind.ClassDeclaration);
+                {
+                    N(SyntaxKind.ClassKeyword);
+                    N(SyntaxKind.IdentifierToken);
+                    N(SyntaxKind.OpenBraceToken);
+                    N(SyntaxKind.IncompleteMember);
+                    {
+                        N(SyntaxKind.ReplaceKeyword);
+                        N(SyntaxKind.GenericName);
+                        {
+                            N(SyntaxKind.IdentifierToken);
+                            N(SyntaxKind.TypeArgumentList);
+                            {
+                                N(SyntaxKind.LessThanToken);
+                                N(SyntaxKind.IdentifierName);
+                                {
+                                    N(SyntaxKind.IdentifierToken);
+                                }
+                                N(SyntaxKind.GreaterThanToken);
+                            }
+                        }
+                    }
+                    N(SyntaxKind.CloseBraceToken);
+                }
+                N(SyntaxKind.EndOfFileToken);
+            }
+        }
+
+        [WorkItem(609912, "http://vstfdevdiv:8080/DevDiv2/DevDiv/_workitems/edit/609912")]
+        [Fact]
+        public void IncompleteReplaceMember06()
+        {
+            // when parsing an incomplete member, treat 'replace' as a modifier if it makes sense
+            // negative case
+            UsingTree(@"
+class C
+{
+    replace Task<T> f
+}");
+            N(SyntaxKind.CompilationUnit);
+            {
+                N(SyntaxKind.ClassDeclaration);
+                {
+                    N(SyntaxKind.ClassKeyword);
+                    N(SyntaxKind.IdentifierToken);
+                    N(SyntaxKind.OpenBraceToken);
+                    N(SyntaxKind.FieldDeclaration);
+                    {
+                        N(SyntaxKind.ReplaceKeyword);
+                        N(SyntaxKind.VariableDeclaration);
+                        {
+                            N(SyntaxKind.GenericName);
+                            {
+                                N(SyntaxKind.IdentifierToken);
+                                N(SyntaxKind.TypeArgumentList);
+                                {
+                                    N(SyntaxKind.LessThanToken);
+                                    N(SyntaxKind.IdentifierName);
+                                    {
+                                        N(SyntaxKind.IdentifierToken);
+                                    }
+                                    N(SyntaxKind.GreaterThanToken);
+                                }
+                            }
+                            N(SyntaxKind.VariableDeclarator);
+                            {
+                                N(SyntaxKind.IdentifierToken);
+                            }
+                            M(SyntaxKind.SemicolonToken);
+                        }
+                    }
+                    N(SyntaxKind.CloseBraceToken);
+                }
+                N(SyntaxKind.EndOfFileToken);
+            }
+        }
+
+        [Fact]
+        public void PropertyReplaceReplace()
+        {
+            UsingTree(@"
+class C
+{
+    replace replace { get; set; }
+}
+");
+            N(SyntaxKind.CompilationUnit);
+            {
+                N(SyntaxKind.ClassDeclaration);
+                {
+                    N(SyntaxKind.ClassKeyword);
+                    N(SyntaxKind.IdentifierToken);
+                    N(SyntaxKind.OpenBraceToken);
+                    N(SyntaxKind.PropertyDeclaration);
+                    {
+                        N(SyntaxKind.IdentifierName);
+                        {
+                            N(SyntaxKind.IdentifierToken);
+                        }
+                        N(SyntaxKind.IdentifierToken);
+                        N(SyntaxKind.AccessorList);
+                        {
+                            N(SyntaxKind.OpenBraceToken);
+                            N(SyntaxKind.GetAccessorDeclaration);
+                            {
+                                N(SyntaxKind.GetKeyword);
+                                N(SyntaxKind.SemicolonToken);
+                            }
+                            N(SyntaxKind.SetAccessorDeclaration);
+                            {
+                                N(SyntaxKind.SetKeyword);
+                                N(SyntaxKind.SemicolonToken);
+                            }
+                            N(SyntaxKind.CloseBraceToken);
+                        }
+                    }
+                    N(SyntaxKind.CloseBraceToken);
+                }
+                N(SyntaxKind.EndOfFileToken);
+            }
+        }
+
+        [Fact]
+        public void PropertyReplaceReplaceReplace()
+        {
+            UsingTree(@"
+class C
+{
+    replace replace replace { get; set; }
+}
+");
+            N(SyntaxKind.CompilationUnit);
+            {
+                N(SyntaxKind.ClassDeclaration);
+                {
+                    N(SyntaxKind.ClassKeyword);
+                    N(SyntaxKind.IdentifierToken);
+                    N(SyntaxKind.OpenBraceToken);
+                    N(SyntaxKind.PropertyDeclaration);
+                    {
+                        N(SyntaxKind.ReplaceKeyword);
+                        N(SyntaxKind.IdentifierName);
+                        {
+                            N(SyntaxKind.IdentifierToken);
+                        }
+                        N(SyntaxKind.IdentifierToken);
+                        N(SyntaxKind.AccessorList);
+                        {
+                            N(SyntaxKind.OpenBraceToken);
+                            N(SyntaxKind.GetAccessorDeclaration);
+                            {
+                                N(SyntaxKind.GetKeyword);
+                                N(SyntaxKind.SemicolonToken);
+                            }
+                            N(SyntaxKind.SetAccessorDeclaration);
+                            {
+                                N(SyntaxKind.SetKeyword);
+                                N(SyntaxKind.SemicolonToken);
+                            }
+                            N(SyntaxKind.CloseBraceToken);
+                        }
+                    }
+                    N(SyntaxKind.CloseBraceToken);
+                }
+                N(SyntaxKind.EndOfFileToken);
+            }
+        }
+
+        [Fact]
+        public void EventReplaceReplace()
+        {
+            UsingTree(@"
+class C
+{
+    event replace replace;
+}
+");
+            N(SyntaxKind.CompilationUnit);
+            {
+                N(SyntaxKind.ClassDeclaration);
+                {
+                    N(SyntaxKind.ClassKeyword);
+                    N(SyntaxKind.IdentifierToken);
+                    N(SyntaxKind.OpenBraceToken);
+                    N(SyntaxKind.EventFieldDeclaration);
+                    {
+                        N(SyntaxKind.EventKeyword);
+                        N(SyntaxKind.VariableDeclaration);
+                        {
+                            N(SyntaxKind.IdentifierName);
+                            {
+                                N(SyntaxKind.IdentifierToken);
+                            }
+                            N(SyntaxKind.VariableDeclarator);
+                            {
+                                N(SyntaxKind.IdentifierToken);
+                            }
+                        }
+                        N(SyntaxKind.SemicolonToken);
+                    }
+                    N(SyntaxKind.CloseBraceToken);
+                }
+                N(SyntaxKind.EndOfFileToken);
+            }
+        }
+
+        [Fact]
+        public void EventReplaceReplaceReplace1()
+        {
+            UsingTree(@"
+class C
+{
+    event replace replace replace;
+}
+");
+            N(SyntaxKind.CompilationUnit);
+            {
+                N(SyntaxKind.ClassDeclaration);
+                {
+                    N(SyntaxKind.ClassKeyword);
+                    N(SyntaxKind.IdentifierToken);
+                    N(SyntaxKind.OpenBraceToken);
+                    N(SyntaxKind.EventFieldDeclaration);
+                    {
+                        N(SyntaxKind.EventKeyword);
+                        N(SyntaxKind.VariableDeclaration);
+                        {
+                            N(SyntaxKind.IdentifierName);
+                            {
+                                N(SyntaxKind.IdentifierToken);
+                            }
+                            N(SyntaxKind.VariableDeclarator);
+                            {
+                                N(SyntaxKind.IdentifierToken);
+                            }
+                        }
+                        M(SyntaxKind.SemicolonToken);
+                    }
+                    N(SyntaxKind.IncompleteMember);
+                    {
+                        N(SyntaxKind.IdentifierName);
+                        {
+                            N(SyntaxKind.IdentifierToken);
+                        }
+                    }
+                    N(SyntaxKind.CloseBraceToken);
+                }
+                N(SyntaxKind.EndOfFileToken);
+            }
+        }
+
+        [Fact]
+        public void EventReplaceReplaceReplace2()
+        {
+            UsingTree(@"
+class C
+{
+    replace event replace replace;
+}
+");
+            N(SyntaxKind.CompilationUnit);
+            {
+                N(SyntaxKind.ClassDeclaration);
+                {
+                    N(SyntaxKind.ClassKeyword);
+                    N(SyntaxKind.IdentifierToken);
+                    N(SyntaxKind.OpenBraceToken);
+                    N(SyntaxKind.EventFieldDeclaration);
+                    {
+                        N(SyntaxKind.ReplaceKeyword);
+                        N(SyntaxKind.EventKeyword);
+                        N(SyntaxKind.VariableDeclaration);
+                        {
+                            N(SyntaxKind.IdentifierName);
+                            {
+                                N(SyntaxKind.IdentifierToken);
+                            }
+                            N(SyntaxKind.VariableDeclarator);
+                            {
+                                N(SyntaxKind.IdentifierToken);
+                            }
+                        }
+                        N(SyntaxKind.SemicolonToken);
+                    }
+                    N(SyntaxKind.CloseBraceToken);
+                }
+                N(SyntaxKind.EndOfFileToken);
+            }
+        }
+    }
+}

--- a/src/Compilers/CSharp/Test/Syntax/Parsing/ReplaceParsingTests.cs
+++ b/src/Compilers/CSharp/Test/Syntax/Parsing/ReplaceParsingTests.cs
@@ -4,15 +4,15 @@ using Microsoft.CodeAnalysis.CSharp.Symbols;
 using Microsoft.CodeAnalysis.CSharp.Syntax;
 using Microsoft.CodeAnalysis.CSharp.Test.Utilities;
 using Microsoft.CodeAnalysis.Test.Utilities;
+using Roslyn.Test.Utilities;
 using System.Linq;
 using Xunit;
 
 namespace Microsoft.CodeAnalysis.CSharp.UnitTests
 {
     [CompilerTrait(CompilerFeature.SourceGenerators)]
-    public class ReplaceParsingTests:  CSharpTestBase
+    public class ReplaceParsingTests : CSharpTestBase
     {
-        
         [Fact]
         public void ReplaceClass()
         {
@@ -23,7 +23,6 @@ namespace Microsoft.CodeAnalysis.CSharp.UnitTests
                 type.Modifiers.ToDeclarationModifiers(),
                 DeclarationModifiers.Abstract | DeclarationModifiers.Override | DeclarationModifiers.Replace);
         }
-
         
         [Fact]
         public void ReplaceMethod()
@@ -39,7 +38,6 @@ namespace Microsoft.CodeAnalysis.CSharp.UnitTests
                 method.Modifiers.ToDeclarationModifiers(),
                 DeclarationModifiers.Virtual | DeclarationModifiers.Protected | DeclarationModifiers.Replace);
         }
-
         
         [Fact]
         public void ReplaceMethodNoFeature()
@@ -52,7 +50,6 @@ namespace Microsoft.CodeAnalysis.CSharp.UnitTests
             );
         }
 
-        
         [Fact]
         public void OriginalExpression()
         {
@@ -63,7 +60,6 @@ namespace Microsoft.CodeAnalysis.CSharp.UnitTests
             var opKind = SyntaxFacts.GetInstanceExpression(SyntaxKind.OriginalKeyword);
             Assert.Equal(SyntaxKind.None, opKind);
         }
-
         
         [Fact]
         public void OriginalNoReplace()
@@ -76,7 +72,6 @@ namespace Microsoft.CodeAnalysis.CSharp.UnitTests
             OriginalInMember("class C { object P => original; }", inReplace: false);
             OriginalInMember("class C { object this[int index] => original[index]; }", inReplace: false);
         }
-
         
         [Fact]
         public void OriginalInReplace()
@@ -99,6 +94,14 @@ namespace Microsoft.CodeAnalysis.CSharp.UnitTests
             var expr = token.Parent;
             var expectedKind = inReplace ? SyntaxKind.OriginalExpression : SyntaxKind.IdentifierName;
             Assert.Equal(expectedKind, expr.Kind());
+        }
+
+        [Fact]
+        [WorkItem(11503, "https://github.com/dotnet/roslyn/issues/11503")]
+        public void MethodCalledReplace()
+        {
+            var root = SyntaxFactory.ParseCompilationUnit("class C { public static object replace() { } }", options: TestOptions.Regular.WithReplaceFeature());
+            root.Errors().Verify();
         }
     }
 }

